### ISSUE TITLE
Aerials not adding properly

### ISSUE THREAD

**Issue by AlexHaffer** [FalconLee1011/Customized-Aerial-Screen-Saver#12]
Hi, I love the idea of adding custom Aerials, but for some reason it no longer works for me.

I can add Aerials without issues in the app, but the thumbnail image is not generated for them. In System Settings, they are just an error icon or black rectangle in the thumbnail, which adds up, given there is no thumbnail in the folder. When selected, a Download button shows, but is not doing anything on click (like in [issue #4](https://github.com/FalconLee1011/Customized-Aerial-Screen-Saver/issues/4)) and instead of the newly added Aerial, another screen saver is used (an older custom one that still works).

The video file I used to test is a 40s long MOV file with 1920⨉1080px @ 60fps and no audio track and the thumbnail image is a JPG file with 1920⨉1080px @ 72dps, which both match the suggested formats.

As mentioned before, a previously added Aerial still works fine, so the new one not working made me curious. I fiddled around with the old one and noticed that changing the thumbnail image within the folder does not change the thumbnail in System Settings, even after refreshing the assets and restarting the Mac. Neither does creating and adding a preview image in the folder manually resolve the issues with the newly added Aerial.

I'm using a MacBook Air M2 (2023) with macOS Sonoma 14.5 currently installed.

![aerial_issue](https://github.com/FalconLee1011/Customized-Aerial-Screen-Saver/assets/80551273/45ec8221-71f8-4895-8984-f54cfb523bcd)

**Comment by AlexHaffer** [comment]
**Update:** I just now noticed after looking at the screenshot again that the path was still broken and read _asset-p**er**view_ instead of _asset-preview_ (as in [issue #2](https://github.com/FalconLee1011/Customized-Aerial-Screen-Saver/issues/2)), so I renamed it and refreshed. Now the addition of new Aerials works fine, but they only play once. Every subsequent play (initiated by clicking preview in System Settings or actually starting the screen saver) is just a black screen until the screen saver is de- and re-selected in System Settings. This change sadly also broke my old Aerial (the thumbnail no longer shows in the app and it only plays once, just like the newly added one).

![aerial_issue](https://github.com/FalconLee1011/Customized-Aerial-Screen-Saver/assets/80551273/d8f27fef-dc94-4268-8256-32d5ec1e62b3)

**Comment by D051p0** [comment]
Same here. Sonoma 14.5 playing wallpaper only once and then stops.

**Comment by larryqiann** [comment]
I did some testing.

First download at least two stock wallpapers, open this app, then open the system assets folder. Rename an existing file to another system file (i.e. swap the filenames on two files). Then click Refresh assetsd a few times and try to go to settings again (make sure the "target" wallpapers aren't selected before doing this). Then select them again in settings.

You will be able to observe that the media plays as expected, and the image does not match as they are switched.

Performing this procedure with a different video file that plays normally in QuickTime appears to result in the bug mentioned above. Thus, I assume that the bug has to do with the format or encoding of the video. I have tried to use Claude to clone the settings and reencode my video similarly, however, this has the same bug. 

This is the output of ffprobe on the original video:

`ffprobe version 7.0.1 Copyright (c) 2007-2024 the FFmpeg developers
  built with Apple clang version 15.0.0 (clang-1500.3.9.4)
  configuration: --prefix=/opt/homebrew/Cellar/ffmpeg/7.0.1 --enable-shared --enable-pthreads --enable-version3 --cc=clang --host-cflags= --host-ldflags='-Wl,-ld_classic' --enable-ffplay --enable-gnutls --enable-gpl --enable-libaom --enable-libaribb24 --enable-libbluray --enable-libdav1d --enable-libharfbuzz --enable-libjxl --enable-libmp3lame --enable-libopus --enable-librav1e --enable-librist --enable-librubberband --enable-libsnappy --enable-libsrt --enable-libssh --enable-libsvtav1 --enable-libtesseract --enable-libtheora --enable-libvidstab --enable-libvmaf --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libxvid --enable-lzma --enable-libfontconfig --enable-libfreetype --enable-frei0r --enable-libass --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libspeex --enable-libsoxr --enable-libzmq --enable-libzimg --disable-libjack --disable-indev=jack --enable-videotoolbox --enable-audiotoolbox --enable-neon
  libavutil      59.  8.100 / 59.  8.100
  libavcodec     61.  3.100 / 61.  3.100
  libavformat    61.  1.100 / 61.  1.100
  libavdevice    61.  1.100 / 61.  1.100
  libavfilter    10.  1.100 / 10.  1.100
  libswscale      8.  1.100 /  8.  1.100
  libswresample   5.  1.100 /  5.  1.100
  libpostproc    58.  1.100 / 58.  1.100
Input #0, mov,mp4,m4a,3gp,3g2,mj2, from 'DE851E6D-C2BE-4D9F-AB54-0F9CE994DC51.mov':
  Metadata:
    major_brand     : qt  
    minor_version   : 0
    compatible_brands: qt  
    creation_time   : 2023-07-19T15:27:52.000000Z
  Duration: 00:04:00.24, start: 0.000000, bitrate: 9239 kb/s
  Stream #0:0[0x1](und): Video: hevc (Main 10) (hvc1 / 0x31637668), yuv420p10le(tv, bt709), 3840x2160 [SAR 1:1 DAR 16:9], 9214 kb/s, 239.76 fps, 239.76 tbr, 239760 tbn (default)
      Metadata:
        creation_time   : 2023-07-19T15:27:52.000000Z
        handler_name    : Core Media Video
        vendor_id       : [0][0][0][0]
        encoder         : HEVC
`

**Comment by larryqiann** [comment]
Update: I managed to try using Apple Compressor to compress the video (because I assumed this is what Apple mastered them with) and there are really not that many settings in there to change to try to encode this properly. Both with frame reordering on and off, the videos play once and then go into a black screen.

Unfortunately at this point it's unclear what exactly causes the issue - looking in the Console it's just giving generic "failed to decode frame" errors which doesn't make sense because it's totally capable of playing the video the first time. 

Of course the same videos in question simply don't work at all in Sequoia when I tried them so this isn't exactly something I think the new OS will fix. It did work correctly on Sonoma before so it was definitely a software update that broke this (unsure if on purpose). Does anyone know whether there are specific tools to see what's going on in the original video files, or if there is some data in there we don't know about?

**Comment by likebliss** [comment]
Having the same issue.  I noticed that the original .mov files are 240 fps.  Could that have something to do with it not working?

**Comment by larryqiann** [comment]
> Having the same issue. I noticed that the original .mov files are 240 fps. Could that have something to do with it not working?

This is not likely the problem. There is something very particular with the encoding of these files. 

I have tried to replicate exactly the encoding settings, profile, etc of these videos using ffmpeg. I suspect that there is a particular choice of I frame interval that is necessary. We do know that some iPhone videos is closer to working than something encoded with ffmpeg, but I don't know what about these videos makes them behave slightly better.

**Comment by Proton0** [comment]
I found out that by killing `WallpaperVideoExtension` every time you log-in fixes it

I made a simple app which does it automatically called [WallpaperVideoExtensionFix](https://github.com/Proton0/WallpaperVideoExtensionFix)

**Comment by larryqiann** [comment]
At this point I'm just curious what it is about the original Apple videos that make them so special.

**Comment by Proton0** [comment]
> At this point I'm just curious what it is about the original Apple videos that make them so special. 

I think its because they are slow-mo videos (indicated by the bar in QuickTime)

**Comment by KiwiJanus** [comment]
I would agree it's perhaps related to the fact that the videos are some type of slow motion. Using a normal wallpaper, once you login, it progressively slows to a stop. Perhaps it's modifying the speed of the clip using a method unique to slow-mo videos. 

I've spent a lot of time using exiftool, ffmpeg, and compressor to almost perfectly match times, framerates, and metadata, but I'm always seeing those same errors in console as well.

You can easily rename a default video wallpaper to match another, and that works fine, meaning it's not related to the length, size, and creation/modify dates of the files either.

**Comment by Proton0** [comment]
Maybe we should de-compile the aerial services to find out

<img width="270" alt="image" src="https://github.com/user-attachments/assets/0f4e6b2e-2c51-4ee4-a5a3-07933ad3b828" />

The files are at `/System/Library/ExtensionKit/Extensions`

**Comment by KawaiiZapic** [comment]
> > At this point I'm just curious what it is about the original Apple videos that make them so special.
> 
> I think its because they are slow-mo videos (indicated by the bar in QuickTime)

No, it doesn’t matter. The bar in QuickTime appears when video frame rate is greater or equal to 120

**Comment by yutaniw** [comment]
i tried all the solutions, but nothing worked .
i found an app that supports the lock screen: https://cindori.com/backdrop
it doesn't have the black screen issue. so maybe it can work ?

**Comment by elliotcz97** [comment]
> i tried all the solutions, but nothing worked . i found an app that supports the lock screen: https://cindori.com/backdrop it doesn't have the black screen issue. so maybe it can work ?

If you tried the App then you probably noticed that every video you try to use as custom aerial is converted! So I guess it’s only a video encode problem that needs to be solved and then it should work.

**Comment by Proton0** [comment]
unless someone manages to decompile backdrop or the aerial services, i doubt this issue will be fixed

**Comment by Proton0** [comment]
I tried backdrop; it worked but as soon as i stopped and uninstalled backdrop and the aerial started glitching again
